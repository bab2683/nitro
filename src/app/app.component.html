<app-header
  [groupingChoices]="groupingChoices"
  [active]="activeChoice"
  (selected)="activeChoice = $event"
></app-header>
<main class="page">
  <div class="page__content">
    <app-tree-view [data]="treeData$ | async"></app-tree-view>
  </div>
</main>
<app-footer></app-footer>
