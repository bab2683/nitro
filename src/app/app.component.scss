@import "common";

:host {
  display: flex;
  flex-direction: column;
  min-height: 100vh;
}

.page {
  flex-grow: 1;

  &__content {
    margin: calc-spacing(2) auto calc-spacing(3);

    @include max-content-width;
  }
}
