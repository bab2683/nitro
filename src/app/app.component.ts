import { Component, OnInit } from '@angular/core';
import { BehaviorSubject, combineLatest, Observable } from 'rxjs';
import { map } from 'rxjs/operators';

import { ParsedPost, Post, TreeData } from '@models';
import { RequestService } from '@services';
import { createSummary, dateDataFromString } from '@utils';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss'],
})
export class AppComponent implements OnInit {
  public title = 'nitro-test';
  public groupingChoices: string[] = ['weekNumber', 'author', 'location'];
  public keyToGroupPosts$: BehaviorSubject<string> = new BehaviorSubject(
    this.groupingChoices[0]
  );
  public posts$: Observable<ParsedPost[]>;
  public treeData$: Observable<TreeData<ParsedPost[]>[]>;

  constructor(private req: RequestService) {}

  public get activeChoice() {
    return this.keyToGroupPosts$.getValue();
  }

  public set activeChoice(value: string) {
    this.keyToGroupPosts$.next(value);
  }

  public ngOnInit() {
    this.posts$ = this.req
      .get<Post[]>({ url: '/api/posts' })
      .pipe(
        map(({ data }) => data),
        map((posts) =>
          posts.map((post) => ({
            ...post,
            ...dateDataFromString(post.time),
            summary: createSummary(post.text, 4),
            open: false,
          }))
        )
      );

    this.treeData$ = combineLatest(this.keyToGroupPosts$, this.posts$).pipe(
      map(([key, posts]) =>
        posts.reduce((result, current) => {
          const objKey: string = `${key}: ${current[key]}`;

          if (result[objKey]) {
            result[objKey].push(this.postToTreeElement(current));
          } else {
            result[objKey] = [this.postToTreeElement(current)];
          }
          return result;
        }, {})
      ),
      map((groupedPosts) =>
        Object.keys(groupedPosts).reduce(
          (result, current) =>
            result.concat([{ title: current, items: groupedPosts[current] }]),
          []
        )
      ),
      map((groupedPosts) => [
        {
          title: 'Posts',
          items: groupedPosts,
        },
      ])
    );
  }

  private postToTreeElement(post: ParsedPost): TreeData<any, ParsedPost> {
    return {
      title: post.summary,
      content: post,
    };
  }
}
