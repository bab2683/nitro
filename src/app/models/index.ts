export * from './date.model';
export * from './post.model';
export * from './request.model';
export * from './tree.model';
