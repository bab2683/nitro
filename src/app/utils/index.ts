export * from './post.utils';
export * from './time.utils';
