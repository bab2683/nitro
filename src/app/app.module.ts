import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { BrowserModule } from '@angular/platform-browser';

import { AppComponent } from './app.component';
import { HeaderComponent } from './components/header/header.component';
import { FooterComponent } from './components/footer/footer.component';
import { PostComponent } from './components/post/post.component';
import { TreeViewComponent } from './components/tree-view/tree-view.component';
import { TreeButtonComponent } from './components/tree-button/tree-button.component';
import { ButtonComponent } from './components/button/button.component';

@NgModule({
  declarations: [
    AppComponent,
    HeaderComponent,
    FooterComponent,
    PostComponent,
    TreeViewComponent,
    TreeButtonComponent,
    ButtonComponent,
  ],
  imports: [
    BrowserModule,
    CommonModule,
    FormsModule,
    HttpClientModule,
    ReactiveFormsModule,
  ],
  bootstrap: [AppComponent],
})
export class AppModule {}
