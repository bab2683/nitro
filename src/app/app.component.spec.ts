import { NO_ERRORS_SCHEMA } from '@angular/compiler/src/core';
import { TestBed, async, ComponentFixture } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { of } from 'rxjs';

import { RequestService } from '@services';
import { AppComponent } from './app.component';

const RequestServiceMock = {
  get: jest.fn().mockReturnValue(
    of({
      data: [
        {
          id: 1,
          location: 'San Francisco',
          time: '1582848000000',
          author: 'Happy User',
          text:
            'Proper PDF conversion ensures that every element of your document remains just as you left it.',
        },
        {
          id: 2,
          location: 'San Francisco',
          time: '1582848000000',
          author: 'Happy User',
          text:
            'Proper PDF conversion ensures that every element of your document remains just as you left it.',
        },
      ],
    })
  ),
};

describe('AppComponent', () => {
  let component: AppComponent;
  let fixture: ComponentFixture<AppComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [RouterTestingModule],
      declarations: [AppComponent],
      schemas: [NO_ERRORS_SCHEMA],
      providers: [
        {
          provide: RequestService,
          useValue: RequestServiceMock,
        },
      ],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(AppComponent);
    component = fixture.componentInstance;
  });

  it('should create the app', () => {
    const fixture = TestBed.createComponent(AppComponent);
    const app = fixture.componentInstance;
    expect(app).toBeTruthy();
  });

  describe('ngOnInit', () => {
    it('should request the posts', () => {
      expect(component.posts$).toBeUndefined();

      component.ngOnInit();

      expect(component.posts$).toBeTruthy();

      expect(RequestServiceMock.get).toHaveBeenCalledWith({
        url: '/api/posts',
      });

      component.posts$.subscribe((data) => {
        expect(data).toMatchSnapshot();
      });

      component.postGroupedByKey$.subscribe((data) => {
        expect(data).toMatchSnapshot();
      });
    });
  });
});
