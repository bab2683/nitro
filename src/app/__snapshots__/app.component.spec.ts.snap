// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent ngOnInit should request the posts 1`] = `
Array [
  Object {
    "author": "Happy User",
    "date": 2020-02-28T00:00:00.000Z,
    "id": 1,
    "location": "San Francisco",
    "open": false,
    "summary": "Proper PDF conversion ensures...",
    "text": "Proper PDF conversion ensures that every element of your document remains just as you left it.",
    "time": "1582848000000",
    "weekNumber": 9,
  },
  Object {
    "author": "Happy User",
    "date": 2020-02-28T00:00:00.000Z,
    "id": 2,
    "location": "San Francisco",
    "open": false,
    "summary": "Proper PDF conversion ensures...",
    "text": "Proper PDF conversion ensures that every element of your document remains just as you left it.",
    "time": "1582848000000",
    "weekNumber": 9,
  },
]
`;

exports[`AppComponent ngOnInit should request the posts 2`] = `
Array [
  Object {
    "items": Array [
      Object {
        "items": Array [
          Object {
            "content": Object {
              "author": "Happy User",
              "date": 2020-02-28T00:00:00.000Z,
              "id": 1,
              "location": "San Francisco",
              "open": false,
              "summary": "Proper PDF conversion ensures...",
              "text": "Proper PDF conversion ensures that every element of your document remains just as you left it.",
              "time": "1582848000000",
              "weekNumber": 9,
            },
            "title": "Proper PDF conversion ensures...",
          },
          Object {
            "content": Object {
              "author": "Happy User",
              "date": 2020-02-28T00:00:00.000Z,
              "id": 2,
              "location": "San Francisco",
              "open": false,
              "summary": "Proper PDF conversion ensures...",
              "text": "Proper PDF conversion ensures that every element of your document remains just as you left it.",
              "time": "1582848000000",
              "weekNumber": 9,
            },
            "title": "Proper PDF conversion ensures...",
          },
        ],
        "title": "weekNumber: 9",
      },
    ],
    "title": "Posts",
  },
]
`;
